libflux/src/semantic/nodes.rs

```diff
@@ -0,0 +1,576 @@
+    Duration(DurationLit),
+    Uint(UintLit),
+    Boolean(BooleanLit),
+    DateTime(DateTimeLit),
+    Regexp(RegexpLit),
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Package {
+    pub loc: ast::SourceLocation,
+
+    pub package: String,
+    pub files: Vec<File>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct File {
+    pub loc: ast::SourceLocation,
+
+    pub package: Option<PackageClause>,
+    pub imports: Vec<ImportDeclaration>,
+    pub body: Vec<Statement>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct PackageClause {
+    pub loc: ast::SourceLocation,
+
+    pub name: Identifier,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ImportDeclaration {
+    pub loc: ast::SourceLocation,
+
+    pub alias: Option<Identifier>,
+    pub path: StringLit,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Block {
+    pub loc: ast::SourceLocation,
+
+    pub body: Vec<Statement>,
+}
+
+impl Block {
+    fn return_statement(&self) -> &ReturnStmt {
+        let len = self.body.len();
+        let last = self
+            .body
+            .get(len - 1)
+            .expect("body must have at least one statement");
+        let last: Option<&ReturnStmt> = match last {
+            Statement::Return(rs) => Some(rs),
+            _ => None,
+        };
+        last.expect("last statement must be a return statement")
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct OptionStmt {
+    pub loc: ast::SourceLocation,
+
+    pub assignment: Assignment,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct BuiltinStmt {
+    pub loc: ast::SourceLocation,
+
+    pub id: Identifier,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct TestStmt {
+    pub loc: ast::SourceLocation,
+
+    pub assignment: VariableAssgn,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ExprStmt {
+    pub loc: ast::SourceLocation,
+
+    pub expression: Expression,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ReturnStmt {
+    pub loc: ast::SourceLocation,
+
+    pub argument: Expression,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct VariableAssgn {
+    pub loc: ast::SourceLocation,
+
+    pub id: Identifier,
+    pub init: Expression,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct MemberAssgn {
+    pub loc: ast::SourceLocation,
+
+    pub member: MemberExpr,
+    pub init: Expression,
+}
+
+#[derive(Derivative)]
+#[derivative(Debug, PartialEq, Clone)]
+pub struct StringExpr {
+    pub loc: ast::SourceLocation,
+    #[derivative(PartialEq = "ignore")]
+    pub typ: types::MonoType,
+
+    pub parts: Vec<StringExprPart>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum StringExprPart {
+    Text(TextPart),
+    Interpolated(InterpolatedPart),
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct TextPart {
+    pub loc: ast::SourceLocation,
+
+    pub value: String,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct InterpolatedPart {
+    pub loc: ast::SourceLocation,
+
+    pub expression: Expression,
+}
```

Comment:
TextPart and InterpolatedPart aren't considered expressions on their own and so they aren't given types. However InterpolatedPart contains an expression. This expression will be constrained to be of type `string` initially. Eventually though it will just be constrained to be of kind `StringFormattable`(insert better name here).

Reply:
Yeah, my assumption was that the expression in the interpolated part will be constrained to work, but in the end it would be turned into a string.